node/node.cpp

```diff
@@ -274,6 +274,50 @@ node::send_signed_message(const bzn::uuid_t& uuid, std::shared_ptr<bzn_envelope>
     }
 }
 
+void
+node::send_maybe_signed_message(const boost::asio::ip::tcp::endpoint& ep, std::shared_ptr<bzn_envelope> msg)
+{
+    if (this->options->get_peer_message_signing())
+    {
+        this->find_session(ep)->send_signed_message(msg);
+    }
+    else
+    {
+        msg->set_sender(this->options->get_uuid());
+        msg->set_swarm_id(this->options->get_swarm_id());
+        this->find_session(ep)->send_message(std::make_shared<std::string>(msg->SerializeAsString()));
+    }
+}
+
+void
+node::multicast_maybe_signed_message(std::shared_ptr<std::vector<boost::asio::ip::tcp::endpoint>> eps, std::shared_ptr<bzn_envelope> msg)
+{
+    this->io_context->post(
+        [self =  shared_from_this(), msg, eps]()
+        {
+            for (const auto& ep : *eps)
+            {
+                self->send_maybe_signed_message(ep, msg);
```

Comment:
This will compute a signature for each peer, even though that signature is identical.

Reply:
The msg is passed as a non-const (shared) pointer down to session::send_signed_message, which signs it if the signature is empty. Shouldn't that do it once and only once?